File: gameplay/player.java
package gameplay;

import java.util.Random;
import java.util.List;
import java.util.ArrayList;

public class player {
    String name;
    int money;
    List<property> propertyList;
    int currentLocation;
    int size;
    int jail;
    boolean jailFreeCard;

    public player(String name, int money, List<property> propertyList, int currentLocation, int size, int jail) {
        this.name = name;
        this.money = money;
        this.propertyList = propertyList;
        this.currentLocation = currentLocation;
        this.size = size;
        this.jail = jail;
        this.jailFreeCard = false;
    }

    public player(String name) {
        this.name = name;
        this.money = 150000;
        this.propertyList = new ArrayList<property>();
        this.currentLocation = 0;
        this.size = 40;
        this.jail = 0;
        this.jailFreeCard = false;
    }

    void setName(String name) {
        this.name = name;
    }

    String getName() {
        return name;
    }

    void addProperty(property p) {
        propertyList.add(p);
    }

    List<property> getPropertyList() {
        return propertyList;
    }

    void addMoney(int m) {
        money += m;
    }

    void subMoney(int m) {
        money -= m;
    }

    int getMoney() {
        return money;
    }

    void updateLocation(int move) {
        currentLocation += move;
        if (currentLocation >= size) {
            System.out.println("You received $200 for passing \"GO!\"");
            money += 200;
        }

        currentLocation = (currentLocation) % size;
    }

    int getLocation() {
        return currentLocation;
    }

    void setLocation(int location) {
        currentLocation = location;
    }

    void setJail() {
        jail = 3;
    }

    void addJailFreeCard() {
        jailFreeCard = true;
    }

    void loseJailFreeCard() {
        jailFreeCard = false;
    }

    void chance(List<player> playerList) {
        Random rand = new Random();
        int chance = rand.nextInt(16) + 1;
        if (chance <= 7) {
            int advance = rand.nextInt(40) + 1;
            updateLocation(advance);
            System.out.println("You advance " + advance + " steps to " + currentLocation + "!");
        } else if (chance == 8) {
            System.out.println("The bank pays you dividends!");
            addMoney(50);
        } else if (chance == 9) {
            System.out.println("You get a Get Out of Jail Free Card!");
            addJailFreeCard();
        } else if (chance == 10) {
            System.out.println("Go back 3 spaces!");
            updateLocation(-3);
        } else if (chance == 11) {
            System.out.println("You go to jail!");
            setLocation(20);
            setJail();
        } else if (chance == 12) {
            System.out.println("You pay repairs for each house you own!");
            for (int i = 0; i < propertyList.size(); i++) {
                int houses = propertyList.get(i).numHouses;
                subMoney(25 * houses);
            }
        } else if (chance == 13) {
            System.out.println("You get a speeding fine!");
            subMoney(15);
        } else if (chance == 14) {
            System.out.println("Move up to nearest NFT!");
            while (!(currentLocation % 5 == 0 && currentLocation % 10 != 0))
                updateLocation(1);
        } else if (chance == 15) {
            System.out.println("You are the chairman. Pay everyone $50 each.");
            for (int i = 0; i < playerList.size(); i++) {
                playerList.get(i).addMoney(50);
                subMoney(50);
            }
        } else {
            System.out.println("Your building loam matures. Collect $150.");
            addMoney(150);
        }

    }

    void community(List<player> playerList) {
        Random rand = new Random();
        int community = rand.nextInt(16);
        if (community == 0) {
            System.out.println("Advance to Go");
            updateLocation(size - currentLocation);
        } else if (community == 1) {
            System.out.println("Bank error in your favor. Collect $200");
            addMoney(200);
        } else if (community == 2) {
            System.out.println("Doctor’s fee. Pay $50");
            subMoney(50);
        } else if (community == 3) {
            System.out.println("From sale of stock you get $50");
            addMoney(50);
        } else if (community == 4) {
            System.out.println("Get Out of Jail Free");
            addJailFreeCard();
        } else if (community == 5) {
            System.out.println("Go to Jail. Go directly to jail, do not pass Go, do not collect $200");
            setLocation(20);
            setJail();
        } else if (community == 6) {
            System.out.println("Holiday fund matures. Receive $100");
            addMoney(100);
        } else if (community == 7) {
            System.out.println("Income tax refund. Collect $20");
            addMoney(20);
        } else if (community == 8) {
            System.out.println("It is your birthday. Collect $10 from every player");
            for (int i = 0; i < playerList.size(); i++) {
                playerList.get(i).subMoney(10);
                addMoney(10);
            }
        } else if (community == 9) {
            System.out.println("Life insurance matures. Collect $100");
            addMoney(100);
        } else if (community == 10) {
            System.out.println("Pay hospital fees of $100");
            subMoney(100);
        } else if (community == 11) {
            System.out.println("Pay school fees of $50");
            subMoney(50);
        } else if (community == 12) {
            System.out.println("Receive $25 consultancy fee");
            addMoney(25);
        } else if (community == 13) {
            System.out.println("You are assessed for street repair. $40 per house.");
            for (int i = 0; i < propertyList.size(); i++) {
                int houses = propertyList.get(i).numHouses;
                addMoney(40 * houses);
            }
        } else if (community == 14) {
            System.out.println("You have won second prize in a beauty contest. Collect $10");
            addMoney(200);
        } else {
            System.out.println("");
            addMoney(10);
        }
    }
}

File: js files/player.js
class player {
  // constructor(name, money, propertyList, currentLocation, size){
  //     this.name = name;
  //     this.money = money;
  //     this.propertyList = propertyList;
  //     this.currentLocation = currentLocation;
  //     this.size = size;
  // }

  constructor(name){
    this.name = name;
    this.money = 150000;
    this.propertyList = [];
    this.currentLocation = 0;
    this.size = 40;
    this.jail = 0;
    this.jailFreeCard = false;
    this.message = [];

    //JS API
    //Create Player Object
    let player = document.createElement( 'li' );
    // player.textContent = '♟️';
    player.textContent = this.name;

    let gridNext = document.querySelector('#grid'+this.currentLocation.toString());
    gridNext.appendChild(player)

    //JS API
    //Create player profile
    this.profile = document.querySelector('#player1');
    this.update();
  }

  update() {
    this.profile.querySelector('.name div').textContent = this.name
    this.profile.querySelector('.jail div').textContent = this.jail
    this.profile.querySelector('.money div').textContent = this.money
    this.profile.querySelector('.location div').textContent = this.currentLocation
  }

  setName(name){
    this.name = name;
  }
  getName(){
    return this.name;
  }

  addProperty(p){
    this.propertyList.push(p);
  }
  getPropertyList(){
    return this.propertyList;
  }

  addMoney(m){
    this.money += m;
  }
  subMoney(m){
    this.money -= m;
  }
  getMoney(){
    return this.money;
  }

  setJail() {
    this.jail = 3;
  }

  addJailFreeCard() {
    this.jailFreeCard = true;
  }

  loseJailFreeCard() {
    this.jailFreeCard = false;
  }

  updateLocation(move){
    let p = this;
    if (p.jail <= 0) {
      //JS API
      //Remove Player Object
      let gridPrevious = document.querySelector('#grid'+this.currentLocation.toString());
      let playerPreviousLocation = gridPrevious.querySelector('li');
      gridPrevious.removeChild(playerPreviousLocation)
      //JS API

      // this.currentLocation = (this.currentLocation + move) % this.size;
      let previousLocation = this.currentLocation;
      // this.currentLocation = this.currentLocation + move;
      let nextLocation = this.currentLocation + move;
      // if (this.currentLocation >= this.size) {
      if (nextLocation >= this.size) {
        this.print("You received $200 for passing \"GO!\"");
        this.money += 200;
      }

      // this.currentLocation = (this.currentLocation) % this.size;
      nextLocation = (nextLocation) % this.size;
      this.moveTo(previousLocation,nextLocation,500);
      console.log("Move to "+this.currentLocation);

      //JS API
      //Create Player Object
      let player = document.createElement( 'li' );
      // player.textContent = '♟️';
      player.textContent = this.name;

      let gridNext = document.querySelector('#grid'+this.currentLocation.toString());
      gridNext.appendChild(player)
      //JS API
    } else {
      p.print(p.name + " is in jail and has to wait " + p.jail + " turns!");
      p.jail = p.jail-1;
    }

  }
  getLocation(){
    return this.currentLocation;
  }

  setLocation(location) {
    //JS API
    //Remove Player Object
    let gridPrevious = document.querySelector('#grid'+this.currentLocation.toString());
    let playerPreviousLocation = gridPrevious.querySelector('li');
    gridPrevious.removeChild(playerPreviousLocation)
    //JS API

    this.currentLocation = location;

    //JS API
    //Create Player Object
    let player = document.createElement( 'li' );
    // player.textContent = '♟️';
    player.textContent = this.name;

    let gridNext = document.querySelector('#grid'+this.currentLocation.toString());
    gridNext.appendChild(player)
    //JS API
  }

  moveTo(previousLocation,location,speed,event) {
    g.isMoving = true;
    let p = this;

    let diff = location - previousLocation;

    if(previousLocation>location)
        diff = previousLocation - location;

    for (let i = 0;i<diff+1; i++) {
      setTimeout(
        function() {

            if(i==diff) {
              p.triggerEvent()
              g.isMoving = false;
              } else {
              let nextLocation = (previousLocation+i+1) % p.size;
            p.setLocation(nextLocation)
              }

        }
        ,speed * i
      )
    }
  }

  print(msg) {
    if(this.message.length == 5){
      this.message.shift();
    }
    this.message.push(msg);
    this.profile.querySelector('.message').innerHTML = "";
    for(let i = 0; i < this.message.length; i++){
      this.profile.querySelector('.message').innerHTML += "<li>"+this.message[i]+"</li>";
    }
  }

  chance(playerList) {
    let chance = Math.floor(Math.random() * (16 + 1 - 0) + 0);
    if (chance <= 7) {
      let advance = Math.floor(Math.random() * (40 + 1 - 0) + 0);;
      this.updateLocation(advance);
      this.print("You advance " + advance + " steps to " + this.currentLocation + "!");
    } else if (chance == 8) {
      this.print("The bank pays you dividends!");
      this.addMoney(50);
    } else if (chance == 9) {
      this.print("You get a Get Out of Jail Free Card!");
      this.addJailFreeCard();
    } else if (chance == 10) {
      this.print("Go back 3 spaces!");
      this.updateLocation(-3);
    } else if (chance == 11) {
      this.print("You go to jail!");
      // this.setLocation(20);
      this.setJail();
      this.moveTo(this.currentLocation,20,100,false)
    } else if (chance == 12) {
      this.print("You pay repairs for each house you own!");
      for (let i = 0; i < propertyList.length; i++) {
        let houses = propertyList[i].numHouses;
        this.subMoney(25 * houses);
      }
    } else if (chance == 13) {
      this.print("You get a speeding fine!");
      this.subMoney(15);
    } else if (chance == 14) {
      this.print("Move up to nearest NFT!");
      while (!(this.currentLocation % 5 == 0 && this.currentLocation % 10 != 0))
      this.updateLocation(1);
    } else if (chance == 15) {
      this.print("You are the chairman. Pay everyone $50 each.");
      for (let i = 0; i < playerList.length; i++) {
        playerList[i].addMoney(50);
        this.subMoney(50);
      }
    } else {
      this.print("Your building loam matures. Collect $150.");
      this.addMoney(150);
    }

  }

  community(playerList) {
    let community = Math.floor(Math.random() * (16 - 0) + 0);
    if (community == 0) {
      this.print("Advance to Go");
      this.updateLocation(this.size - this.currentLocation);
    } else if (community == 1) {
      this.print("Bank error in your favor. Collect $200");
      this.addMoney(200);
    } else if (community == 2) {
      this.print("Doctor’s fee. Pay $50");
      this.subMoney(50);
    } else if (community == 3) {
      this.print("From sale of stock you get $50");
      this.addMoney(50);
    } else if (community == 4) {
      this.print("Get Out of Jail Free");
      this.addJailFreeCard();
    } else if (community == 5) {
      this.print("Go to Jail. Go directly to jail, do not pass Go, do not collect $200");
      // this.setLocation(20);
      this.setJail();
      this.moveTo(this.currentLocation,20,100,false)
    } else if (community == 6) {
      this.print("Holiday fund matures. Receive $100");
      this.addMoney(100);
    } else if (community == 7) {
      this.print("Income tax refund. Collect $20");
      this.addMoney(20);
    } else if (community == 8) {
      this.print("It is your birthday. Collect $10 from every player");
      for (let i = 0; i < playerList.length; i++) {
        playerList[i].subMoney(10);
        this.addMoney(10);
      }
    } else if (community == 9) {
      this.print("Life insurance matures. Collect $100");
      this.addMoney(100);
    } else if (community == 10) {
      this.print("Pay hospital fees of $100");
      this.subMoney(100);
    } else if (community == 11) {
      this.print("Pay school fees of $50");
      this.subMoney(50);
    } else if (community == 12) {
      this.print("Receive $25 consultancy fee");
      this.addMoney(25);
    } else if (community == 13) {
      this.print("You are assessed for street repair. $40 per house.");
      for (let i = 0; i < propertyList.length; i++) {
        let houses = propertyList[i].numHouses;
        this.addMoney(40 * houses);
      }
    } else if (community == 14) {
      this.print("You have won second prize in a beauty contest. Collect $10");
      this.addMoney(200);
    } else {
      this.print("");
      this.addMoney(10);
    }
  }

  addPropertyLevel(currentProperty){
    let currentGrid = document.querySelector('#grid'+this.currentLocation.toString());

    if(currentProperty.location >= 1 && currentProperty.location <= 10){ // upper
      let createDot = document.createElement("div");
      createDot.classList.add("dot");
      // createDot.classList.add("align-self-end");
      currentGrid.appendChild(createDot);
    }
    else if(currentProperty.location >= 11 && currentProperty.location <= 20){ // right side
      let createDot = document.createElement("div");
      createDot.classList.add("dot");
      currentGrid.appendChild(createDot);
    }
    else if(currentProperty.location >= 21 && currentProperty.location <= 30){ // bottom
      let createDot = document.createElement("div");
      createDot.classList.add("dot");
      currentGrid.appendChild(createDot);
    }
    else{ // left
      let createDot = document.createElement("div");
      createDot.classList.add("dot");
      currentGrid.appendChild(createDot);
    }
  }

  triggerEvent() {
    let p = this;
    let currentProperty = g.board[this.getLocation()];
    let type = currentProperty.getType();

    if (type == "chance") {
      p.chance(g.playerList);
      // if (p.jail > 0)
      //     break;
    } else if (type == "gotojail") {
      if (!p.jailFreeCard) {
        p.setJail();
        // p.setLocation(10);
        this.moveTo(this.currentLocation,10,100,false)
      } else
      p.loseJailFreeCard();
    } else if (type == "nft") {
      this.print(p.name + " bought an NFT for $200!");
      p.subMoney(200);
    } else if (type == "tax") {
      this.print("It's tax season!");
      p.subMoney(p.getMoney() / 10);
    } else if (type == "community") {
      p.community(playerList);
    } else if(type == "realestate"){
      if(currentProperty.owned == false){
        if(p.getMoney() >= currentProperty.marketPrice){
          let answer = prompt(`Do you want to buy ${currentProperty.getInfo()} for ${currentProperty.getMarketPrice()}? yes:1, no:0`);
          answer = parseInt(answer);

          if (answer == 1) {
            currentProperty.buyProperty(p);
            p.addProperty(currentProperty);
            alert(`${p.getName()} successfully purchased ${currentProperty.getInfo()}!`);
            this.print(`successfully purchased ${currentProperty.getInfo()}`);
            this.profile.querySelector('.property div').innerHTML += `${currentProperty.getInfo()}`;
            this.addPropertyLevel(currentProperty);
          }
        } else{
          confirm("Insufficient Fund. Cannot buy this property.");
        }
      }
      else{ // if it is owned by someone
        if(currentProperty.getOwner().name == p.name){
          let answer = prompt(`Do you want to upgrade ${currentProperty.getInfo()} for ${currentProperty.getMarketPrice()}? yes:1, no:0`);
          answer = parseInt(answer);
          if (answer == 1) {
            // should have a maximum property level
            if(currentProperty.numHouses === 3){
              alert("Property reached highest level.\nNo further update is allowed");
              return;
            }
            // removed checkColor function!!!!!
            currentProperty.buyProperty(p);
            p.addProperty(currentProperty);
            this.addPropertyLevel(currentProperty);
          }
        }
        else{ // owns by someone else, pay rent
          currentProperty.payRent(p);
        }
      }
    }
  }
}
